File: app/src/main/java/com/allrecipes/presenters/HomeScreenPresenter.java
package com.allrecipes.presenters;

import android.text.TextUtils;
import android.util.Log;

import com.allrecipes.di.managers.FirebaseDatabaseManager;
import com.allrecipes.managers.GoogleYoutubeApiManager;
import com.allrecipes.managers.LocalStorageManagerInterface;
import com.allrecipes.managers.remoteconfig.RemoteConfigManager;
import com.allrecipes.model.DefaultChannel;
import com.allrecipes.model.Channel;
import com.allrecipes.model.FiltersAndSortSettings;
import com.allrecipes.model.RecipeFilterOption;
import com.allrecipes.model.RecommendedPlaylists;
import com.allrecipes.model.SearchChannelVideosResponse;
import com.allrecipes.model.YoutubeItem;
import com.allrecipes.model.YoutubeSnipped;
import com.allrecipes.model.playlist.YoutubeChannelItem;
import com.allrecipes.model.playlist.YoutubePlaylistWithVideos;
import com.allrecipes.model.video.YoutubeVideoResponse;
import com.allrecipes.ui.home.viewholders.items.SwipeLaneChannelItem;
import com.allrecipes.ui.home.views.HomeScreenView;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.lang.ref.WeakReference;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import rx.Subscription;
import rx.functions.Action1;

public class HomeScreenPresenter extends AbstractPresenter<HomeScreenView> {

    private static final String APP_LAST_CHANNEL_USED = "app.lastChannelUsed";

    private final GoogleYoutubeApiManager googleYoutubeApiManager;
    private final LocalStorageManagerInterface localStorageManagerInterface;
    private final FirebaseDatabaseManager firebaseDatabaseManager;
    private final RemoteConfigManager remoteConfigManager;

    private Disposable fetchChannelVideosDisposable;
    private CompositeDisposable disposables = new CompositeDisposable();

    private Channel currentChannel;
    private String pageToken;
    private String oAuthToken;

    public HomeScreenPresenter(
        HomeScreenView view,
        GoogleYoutubeApiManager googleYoutubeApiManager,
        LocalStorageManagerInterface localStorageManagerInterface,
        FirebaseDatabaseManager firebaseDatabaseManager,
        RemoteConfigManager remoteConfigManager
    ) {
        super(new WeakReference<>(view));
        this.googleYoutubeApiManager = googleYoutubeApiManager;
        this.localStorageManagerInterface = localStorageManagerInterface;
        this.firebaseDatabaseManager = firebaseDatabaseManager;
        this.remoteConfigManager = remoteConfigManager;
    }

    @Override
    public void unbindAll() {
        dispose(fetchChannelVideosDisposable);
        disposables.clear();
    }

    public void onChannelListClick(Channel channel, FiltersAndSortSettings currentFilterSettings) {
        saveLastUsedChannel(channel);
        this.currentChannel = channel;
        fetchYoutubeChannelVideos(null, "", currentFilterSettings);
    }

    private void saveLastUsedChannel(Channel channel) {
        String categoryJson = new GsonBuilder().create().toJson(channel, Channel.class);
        localStorageManagerInterface.putString(APP_LAST_CHANNEL_USED, categoryJson);
    }

    private String constructSearchFromFilters(String searchCriteria, List<RecipeFilterOption> filters) {
        StringBuilder filtersCombined = new StringBuilder();

        if (!TextUtils.isEmpty(searchCriteria)) {
            filtersCombined.append(searchCriteria);
            if (!filters.isEmpty()) {
                filtersCombined.append("&");
            }
        }
        for (int i = 0; i < filters.size(); i++) {
            if (filters.get(i).isChecked()) {
                filtersCombined.append(filters.get(i).getRecipeFilter());
                if (i < filters.size() - 1) {
                    filtersCombined.append("|");
                }
            }
        }

        return filtersCombined.toString();
    }

    public void fetchYoutubeChannelVideos(
        final String currentPageToken,
        String searchCriteria,
        FiltersAndSortSettings currentFilterSettings
    ) {
        if (currentPageToken == null) {
            getView().showLoading();
        }
        if (TextUtils.isEmpty(searchCriteria) && TextUtils.isEmpty(currentPageToken)) {
            loadRecommendedPlayLists(currentChannel);
        }

        fetchChannelVideosDisposable = googleYoutubeApiManager
            .fetchChannelVideos(
                currentChannel.getChannelId(),
                currentPageToken,
                remoteConfigManager.getVideoListItemsPerPage(),
                currentFilterSettings.getSort(),
                constructSearchFromFilters(searchCriteria, currentFilterSettings.getFilters()),
                oAuthToken
            )
            .subscribe(new Consumer<SearchChannelVideosResponse>() {
                @Override
                public void accept(
                    @NonNull SearchChannelVideosResponse searchChannelVideosResponse
                ) throws Exception {
                    if (isViewAvailable()) {
                        if (currentPageToken == null) {
                            getView().clearAdapterItems();
                        }
                        getView().removeBottomListProgress();
                        List<YoutubeItem> items = searchChannelVideosResponse.items;
                        int position = 0;
                        for (YoutubeItem item : items) {
                            getView().addYoutubeItemToAdapter(item, position);
                            position++;
                        }
                        pageToken = searchChannelVideosResponse.nextPageToken;
                        getView().hideLoading();
                    }
                }
            }, new Consumer<Throwable>() {
                @Override
                public void accept(@NonNull Throwable throwable) throws Exception {
                    if (isViewAvailable()) {
                        getView().hideLoading();
                        throwable.printStackTrace();
                    }
                }
            });

    }

    public void onLoadMore(String searchCriteria, FiltersAndSortSettings currentFilterSettings) {
        fetchYoutubeChannelVideos(pageToken, searchCriteria, currentFilterSettings);
    }

    private void fetchVideosFromPlaylist(final String channelName,final RecommendedPlaylists recommendedPlaylists) {
       Disposable d = googleYoutubeApiManager.fetchVideosInPlaylist(recommendedPlaylists.getChannelId(),
                                                                    remoteConfigManager.getVideoListItemsPerPage()
           , null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<YoutubeVideoResponse>() {
                    @Override
                    public void accept(@NonNull YoutubeVideoResponse youtubeVideoResponse) throws Exception {
                        YoutubeChannelItem channelItem = new YoutubeChannelItem();
                        YoutubeSnipped youtubeSnipped = new YoutubeSnipped();
                        youtubeSnipped.title = channelName;
                        youtubeSnipped.channelId = recommendedPlaylists.getChannelId();
                        youtubeSnipped.channelTitle = channelName;
                        channelItem.setSnippet(youtubeSnipped);
                        getView().addSwapLaneChannelItemToAdapter(new YoutubePlaylistWithVideos(channelItem, youtubeVideoResponse),recommendedPlaylists.getPosition());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(@NonNull Throwable throwable) throws Exception {
                        Log.e("fetchVideosFromPlaylist","",throwable);
                    }
                });
        disposables.add(d);
    }

    private Observable<YoutubePlaylistWithVideos> getVideosForEachPlaylist(YoutubeChannelItem channel) {
        Observable<YoutubeVideoResponse> fetchVideosInPlaylistObservable = googleYoutubeApiManager.fetchVideosInPlaylist(channel.getId(), 50, null);
        return Observable.zip(Observable.just(channel), fetchVideosInPlaylistObservable,
            new BiFunction<YoutubeChannelItem, YoutubeVideoResponse, YoutubePlaylistWithVideos>() {
                @Override
                public YoutubePlaylistWithVideos apply(@NonNull YoutubeChannelItem youtubeChannelItem, @NonNull YoutubeVideoResponse youtubeVideoResponse) throws Exception {
                    return new YoutubePlaylistWithVideos(youtubeChannelItem, youtubeVideoResponse);
                }
            });
    }

    public void onCreate(String oAuthToken) {
        this.oAuthToken = oAuthToken;

        String lastUsedChannel = localStorageManagerInterface.getString(APP_LAST_CHANNEL_USED, "");
        currentChannel = new GsonBuilder().create().fromJson(lastUsedChannel, Channel.class);

        getView().setToolbarTitleText(currentChannel.getName());
        FiltersAndSortSettings filtersAndSortSettings = initDefaultFilterAndSortSettings();
        fetchYoutubeChannelVideos(null, "", filtersAndSortSettings);

        getView().initChannelsListOverlayAdapter(firebaseDatabaseManager.restoreFirebaseConfig(), 0);
    }

    private void loadRecommendedPlayLists(Channel channel) {
        Map<String, RecommendedPlaylists> recommendedPlayLists = channel.getRecommendedPlaylists();
        for (Map.Entry<String,RecommendedPlaylists> recommended : recommendedPlayLists.entrySet()) {
            if (recommended.getValue().getVisible()) {
                fetchVideosFromPlaylist(recommended.getKey(), recommended.getValue());
            }
        }
    }

    private FiltersAndSortSettings initDefaultFilterAndSortSettings() {
        FiltersAndSortSettings filtersAndSortSettings = new FiltersAndSortSettings();
        filtersAndSortSettings.setSort(remoteConfigManager.getDefaultFilterSort());
        List<String> remoteFilterCategories = remoteConfigManager.getFilterCategories();
        if (remoteFilterCategories != null) {
            for (String filterCategory : remoteFilterCategories) {
                filtersAndSortSettings.getFilters().add(new RecipeFilterOption(filterCategory, false));
            }
        }
        getView().setCurrentFilterSettings(filtersAndSortSettings);

        return filtersAndSortSettings;
    }

    public void fetchMoreVideosFromPlaylist(SwipeLaneChannelItem item) {
        final WeakReference<SwipeLaneChannelItem> weakViewReference = new WeakReference<>(item);
        Disposable d = googleYoutubeApiManager.fetchVideosInPlaylist(item.getItem().getChannel().getSnippet().channelId,
                                                                     remoteConfigManager.getVideoListItemsPerPage(),
                                                                     item.getItem().getVideosResponse().nextPageToken)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(new Consumer<YoutubeVideoResponse>() {
                @Override
                public void accept(@NonNull YoutubeVideoResponse youtubeVideoResponse) throws Exception {
                    SwipeLaneChannelItem view = weakViewReference.get();
                    if (view != null) {
                        view.loadMore(youtubeVideoResponse);
                    }
                }
            }, new Consumer<Throwable>() {
                @Override
                public void accept(@NonNull Throwable throwable) throws Exception {
                    Log.e("fetchVideosFromPlaylist", "", throwable);
                }
            });
        disposables.add(d);
    }
}

File: app/src/main/java/com/allrecipes/presenters/HomeScreenPresenter.kt
package com.allrecipes.presenters

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import com.allrecipes.managers.FavoritesManager
import com.allrecipes.managers.FirebaseDatabaseManager
import com.allrecipes.managers.GoogleYoutubeApiManager
import com.allrecipes.managers.LocalStorageManagerInterface
import com.allrecipes.managers.remoteconfig.RemoteConfigManager
import com.allrecipes.model.*
import com.allrecipes.model.playlist.YoutubeChannelItem
import com.allrecipes.model.playlist.YoutubePlaylistWithVideos
import com.allrecipes.tracking.providers.firebase.FirebaseTracker
import com.allrecipes.ui.home.viewholders.items.SwipeLaneChannelItem
import com.allrecipes.ui.home.views.HomeScreenView
import com.allrecipes.util.Constants
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.gson.GsonBuilder
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function3
import io.reactivex.schedulers.Schedulers
import java.lang.ref.WeakReference

class HomeScreenPresenter(
    view: WeakReference<HomeScreenView>,
    private val googleYoutubeApiManager: GoogleYoutubeApiManager,
    private val localStorageManagerInterface: LocalStorageManagerInterface,
    private val firebaseDatabaseManager: FirebaseDatabaseManager,
    private val remoteConfigManager: RemoteConfigManager,
    private val firebaseTracker: FirebaseTracker,
    private val favoritesManager: FavoritesManager
) : AbstractPresenter<HomeScreenView>(view) {

    private var fetchChannelVideosDisposable: Disposable? = null
    private val disposables = CompositeDisposable()

    private var currentChannel: Channel = Channel()
    private var pageToken: String? = null
    private var favoritesCount = 0

    override fun unbindAll() {
        dispose(fetchChannelVideosDisposable)
        disposables.clear()
    }

    fun onChannelListClick(channel: Channel, currentFilterSettings: FiltersAndSortSettings) {
        saveLastUsedChannel(channel)
        this.currentChannel = channel
        fetchYoutubeChannelVideos(null, "", currentFilterSettings)
    }

    private fun saveLastUsedChannel(channel: Channel) {
        val categoryJson = GsonBuilder().create().toJson(channel, Channel::class.java)
        localStorageManagerInterface.putString(APP_LAST_CHANNEL_USED, categoryJson)
    }

    private fun constructSearchFromFilters(searchCriteria: String?, filters: List<RecipeFilterOption>): String {
        val filtersCombined = StringBuilder()

        if (!TextUtils.isEmpty(searchCriteria)) {
            filtersCombined.append(searchCriteria)
            if (!filters.isEmpty()) {
                filtersCombined.append("&")
            }
        }
        for (i in filters.indices) {
            if (filters[i].isChecked) {
                filtersCombined.append(filters[i].recipeFilter)
                if (i < filters.size - 1) {
                    filtersCombined.append("|")
                }
            }
        }
        if (filtersCombined.contains("|")) {
            filtersCombined.replace(filtersCombined.length - 1, filtersCombined.length, "")
        }

        return filtersCombined.toString()
    }

    fun trackVideoSearch(searchCriteria: String?, sort: String, filters: String) {
        val bundle = Bundle()
        bundle.putString(FirebaseAnalytics.Param.SEARCH_TERM, searchCriteria)
        bundle.putString("filters", filters)
        bundle.putString("sort", sort)
        bundle.putString("channel_id", currentChannel.channelId)
        bundle.putString("channel_name", currentChannel.name)

        firebaseTracker.logEvent(FirebaseAnalytics.Event.SEARCH, bundle)
    }

    fun fetchYoutubeChannelVideos(
        currentPageToken: String?,
        searchCriteria: String?,
        currentFilterSettings: FiltersAndSortSettings
    ) {
        if (currentPageToken == null) {
            getView().showLoading()
            trackVideoSearch(searchCriteria, currentFilterSettings.sort, currentFilterSettings.filters.joinToString())
        }

        if (TextUtils.isEmpty(currentPageToken) && TextUtils.isEmpty(searchCriteria) && !currentFilterSettings.isAtLeastOneFilterSet ) {
            fetchHomeYoutubeVideos(currentPageToken, searchCriteria, currentFilterSettings)
        } else {
            searchYoutubeVideos(currentPageToken, currentFilterSettings, searchCriteria)
        }
    }

    private fun fetchHomeYoutubeVideos(
        currentPageToken: String?,
        searchCriteria: String?,
        currentFilterSettings: FiltersAndSortSettings
    ) {
        val videosObservable: Observable<SearchChannelVideosResponse> = googleYoutubeApiManager
            .fetchChannelVideos(
                currentChannel.channelId,
                currentPageToken,
                remoteConfigManager.videoListItemsPerPage,
                currentFilterSettings.sort,
                constructSearchFromFilters(searchCriteria, currentFilterSettings.filters))

        val swipeLanesObservable: Observable<List<YoutubePlaylistWithVideos>> = loadRecommendedPlayListsZip(currentChannel)
        val favoritesObservable: Observable<YoutubePlaylistWithVideos> = favoritesManager.getFavoriteVideos(currentChannel.channelId)

        Observable.zip(videosObservable, swipeLanesObservable, favoritesObservable,
                       Function3 { youtubeVideos: SearchChannelVideosResponse,
                                   swipeLanes: List<YoutubePlaylistWithVideos>,
                                   favoriteVideos: YoutubePlaylistWithVideos ->
                           val homeZipItems: HomeZipResult = HomeZipResult()
                           homeZipItems.videos = youtubeVideos
                           homeZipItems.swipeLanes = swipeLanes
                           homeZipItems.favoriteVideos = favoriteVideos
                           favoritesCount = favoriteVideos?.videosResponse?.items?.size!!
                           homeZipItems
                       })
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ homeZip ->
                           if (isViewAvailable) {
                               getView().hideLoading()

                               if (currentPageToken == null) {
                                   getView().clearAdapterItems()
                               }
                               getView().removeBottomListProgress()
                               val items = homeZip.videos.items
                               var position = 0
                               items.forEach {
                                   getView().addYoutubeItemToAdapter(it, position)
                                   position++
                               }
                               pageToken = homeZip.videos.nextPageToken

                               homeZip.swipeLanes.forEach {
                                   getView().addSwapLaneChannelItemToAdapter(it, it.position)
                               }

                               val userHasFavVideos: Boolean? = homeZip.favoriteVideos?.videosResponse?.items?.isEmpty()
                               if (userHasFavVideos != null && userHasFavVideos == false) {
                                   getView().addSwapLaneChannelItemToAdapter(homeZip.favoriteVideos!!, 0)
                               }
                           }
                       }, { t ->
                           if (isViewAvailable) {
                               getView().hideLoading()
                               t.printStackTrace()
                               getView().handleApiError(
                                   t, {
                                   fetchYoutubeChannelVideos(
                                       currentPageToken,
                                       searchCriteria,
                                       currentFilterSettings
                                   )
                               })
                           }
                       })
    }

    private fun searchYoutubeVideos(
        currentPageToken: String?,
        currentFilterSettings: FiltersAndSortSettings,
        searchCriteria: String?
    ) {
        fetchChannelVideosDisposable = googleYoutubeApiManager
            .fetchChannelVideos(
                currentChannel.channelId,
                currentPageToken,
                remoteConfigManager.videoListItemsPerPage,
                currentFilterSettings.sort,
                constructSearchFromFilters(searchCriteria, currentFilterSettings.filters)
            )
            .subscribe({ searchChannelVideosResponse ->
                       if (isViewAvailable) {
                           if (currentPageToken == null) {
                               getView().clearAdapterItems()
                           }
                           getView().removeBottomListProgress()
                           val items = searchChannelVideosResponse.items
                           var position = 0
                           items.forEach {
                               getView().addYoutubeItemToAdapter(it, position)
                               position++
                           }
                           pageToken = searchChannelVideosResponse.nextPageToken
                           getView().hideLoading()
                       }
                   }) { throwable ->
                if (isViewAvailable) {
                    getView().hideLoading()
                    throwable.printStackTrace()
                    getView().handleApiError(
                        throwable,
                        { fetchYoutubeChannelVideos(currentPageToken, searchCriteria, currentFilterSettings) }
                    )
                }
            }
    }

    fun onLoadMore(searchCriteria: String?, currentFilterSettings: FiltersAndSortSettings) {
        fetchYoutubeChannelVideos(pageToken, searchCriteria, currentFilterSettings)
    }

    private fun getSwipeLaneChannelSubscription(channelName: String, recommendedPlaylists: RecommendedPlaylists)
        : Observable<YoutubePlaylistWithVideos> {
        val s: Observable<YoutubePlaylistWithVideos> = googleYoutubeApiManager.fetchVideosInPlaylist(
            recommendedPlaylists.playlistId,
            remoteConfigManager.videoListItemsPerPage, null
        )
            .flatMap({ youtubeVideoResponse ->
                         val channelItem = YoutubeChannelItem()
                         val youtubeSnipped = YoutubeSnipped()
                         youtubeSnipped.title = channelName
                         youtubeSnipped.channelId = recommendedPlaylists.playlistId
                         youtubeSnipped.channelTitle = channelName
                         channelItem.snippet = youtubeSnipped
                         val response = YoutubePlaylistWithVideos(
                             channelItem,
                             youtubeVideoResponse,
                             recommendedPlaylists.position
                         )
                         Observable.just(response)
                     })

        return s
    }


    fun onCreate(oAuthToken: String?) {
        if (!TextUtils.isEmpty(oAuthToken)) {
            localStorageManagerInterface.putString(Constants.GOOGLE_LOGIN_ACCESS_TOKEN, oAuthToken)
        }

        val lastUsedChannel = localStorageManagerInterface.getString(APP_LAST_CHANNEL_USED, "")
        currentChannel = GsonBuilder().create().fromJson(lastUsedChannel, Channel::class.java)

        getView().setToolbarTitleText(currentChannel.name)
        val filtersAndSortSettings = initDefaultFilterAndSortSettings()
        fetchYoutubeChannelVideos(null, "", filtersAndSortSettings)

        getView().initChannelsListOverlayAdapter(firebaseDatabaseManager.restoreFirebaseConfig(), 0)
    }

    private fun loadRecommendedPlayListsZip(channel: Channel): Observable<List<YoutubePlaylistWithVideos>> {
        val swipeLanesObservables: ArrayList<Observable<YoutubePlaylistWithVideos>> = ArrayList()
        val recommendedPlayLists = channel.recommendedPlaylists
        for ((key, value) in recommendedPlayLists) {
            if (value.visible) {
                val observable: Observable<YoutubePlaylistWithVideos> = getSwipeLaneChannelSubscription(key, value)
                swipeLanesObservables.add(observable)
            }
        }

        if (swipeLanesObservables.isEmpty()) {
            return Observable.just(emptyList())
        }

        return Observable.zip(swipeLanesObservables, {
            t: Array<out Any> ->
            t.toList() as List<YoutubePlaylistWithVideos>
        })
    }

    private fun initDefaultFilterAndSortSettings(): FiltersAndSortSettings {
        val filtersAndSortSettings = FiltersAndSortSettings()
        filtersAndSortSettings.sort = remoteConfigManager.defaultFilterSort
        val remoteFilterCategories = remoteConfigManager.filterCategories
        if (remoteFilterCategories != null) {
            for (filterCategory in remoteFilterCategories) {
                filtersAndSortSettings.filters.add(RecipeFilterOption(filterCategory, false))
            }
        }
        getView().setCurrentFilterSettings(filtersAndSortSettings)

        return filtersAndSortSettings
    }

    fun fetchMoreVideosFromPlaylist(item: SwipeLaneChannelItem) {
        val weakViewReference = WeakReference(item)
        val d = googleYoutubeApiManager.fetchVideosInPlaylist(
            item.item.channel.snippet!!.channelId,
            remoteConfigManager.videoListItemsPerPage,
            item.item.videosResponse.nextPageToken
        )
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ youtubeVideoResponse ->
                           val view = weakViewReference.get()
                           view?.loadMore(youtubeVideoResponse)
                       }) { throwable -> Log.e("fetchVideosFromPlaylist", "", throwable) }
        disposables.add(d)
    }

    companion object {
        private val APP_LAST_CHANNEL_USED = "app.lastChannelUsed"
    }

    fun onResume(currentFilterSettings: FiltersAndSortSettings) {
        if (favoritesCount != favoritesManager.getChannelFavorites(currentChannel.channelId).size) {
            fetchYoutubeChannelVideos(null, "", currentFilterSettings)
        }
    }
}
